### PR TITLE
Added `using` operators

### LINKED ISSUE #673: Add `using` operators
Docs: https://reactivex.io/documentation/operators/using.html

### PR DESCRIPTION
**Pull request by arkivanov**
Closes #673 

cc: @CherryPerry

### CODE REVIEW

File: reaktive/src/commonTest/kotlin/com/badoo/reaktive/completable/UsingTest.kt [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+            }
+        )
+
+        upstream.onComplete()
+
+        assertFalse(isResourceDisposedBeforeInnerCompletable)
+    }
+
+    @Test
+    fun resource_is_not_disposed_WHEN_eager_is_false_and_downstream_received_onError() {
```
Comment (CherryPerry) [inline reaktive/src/commonTest/kotlin/com/badoo/reaktive/completable/UsingTest.kt:225]: Weird test name, it is hard to understand what you are testing here. `using` by design should invoke `resourceCleanup` but test name says it should not.

```kotlin
@Test
fun resource_is_not_disposed _BEFORE_DOWNSTREAM_ WHEN_eager_is_false_and_downstream_received_onError()
```

File: reaktive/src/commonTest/kotlin/com/badoo/reaktive/completable/UsingTest.kt [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+import com.badoo.reaktive.utils.atomic.AtomicList
+import com.badoo.reaktive.utils.atomic.getValue
+import com.badoo.reaktive.utils.atomic.plusAssign
+import com.badoo.reaktive.utils.atomic.setValue
+import kotlin.test.Test
+import kotlin.test.assertContentEquals
+import kotlin.test.assertFalse
+import kotlin.test.assertTrue
+
+class UsingTest :
```
Comment (CherryPerry) [inline reaktive/src/commonTest/kotlin/com/badoo/reaktive/completable/UsingTest.kt:18]: You can create `class DisposablesFactoryStub: () -> Disposable` that will store all created objects that you can verify later. By doing this we can verify that we invoke `resourceSupplier` only once (or amount of times we needed). Also we can verify that we do not create resource when it is not required.

```kotlin
    @Test
    fun acquires_new_resource_each_time_WHEN_eager_is_true_and_subscribed_multiple_times() {
        val factory = DisposablesFactoryStub()
        val acquiredResources = AtomicList<Disposable>(emptyList())

        val downstream =
            completableUsing(resourceSupplier = factory, eager = true) {
                acquiredResources += it
                TestCompletable()
            }

        repeat(3) { downstream.test() }

        assertContentEquals(factory.created, acquiredResources.value)
    }
```

File: reaktive/src/commonTest/kotlin/com/badoo/reaktive/completable/UsingTest.kt (hunk 1 above)
Comment (arkivanov) [reply reaktive/src/commonTest/kotlin/com/badoo/reaktive/completable/UsingTest.kt:225]: There are two modes: `eager=true` and `eager=false`. The former should dispose the resource before terminating the downstream, the latter should dispose the resource after terminating the downstream. So, when the downstream receives either `onComplete` or `onError`, the resource should already be disposed if `eager=true` and should be yet disposed if `eager=false`.

This test says - resource is not disposed WHEN downstream received `onError`. Perhaps a better name would be "resource_is_not_yet_disposed_WHEN_eager_is_false_and_downstream_received_onError`.

Comment (CherryPerry) [reply reaktive/src/commonTest/kotlin/com/badoo/reaktive/completable/UsingTest.kt:225]: What do you think about `resource_is_not_disposed_BEFORE_DOWNSTREAM_`?

Comment (arkivanov) [reply reaktive/src/commonTest/kotlin/com/badoo/reaktive/completable/UsingTest.kt]: I have updated all tests, now it says `disposes_resource_after_downstream_signalled_onError_WHEN_eager_false_and_upstream_produced_error`.

File: reaktive/src/commonTest/kotlin/com/badoo/reaktive/completable/UsingTest.kt (hunk 2 above)
Comment (arkivanov) [reply reaktive/src/commonTest/kotlin/com/badoo/reaktive/completable/UsingTest.kt:16]: Fixed.

**Review by CherryPerry — APPROVED** [review]

**Pull request merged**